

#' Function to split inputted parental genotypes into alleles for each locus in
#' a diploid species.
#'
#' @param x is a character or string object of genotype for at least one locus.
#' x should be either lower or uppercase letters of the Latin/English alphabet.
#' Uppercase letters represent dominant alleles, and lowercase letters denote
#' recessive alleles.
#'
#' @returns a list object of two elements representing the number of loci and
#' split loci and loci alleles
#'
#' @examples
#'
#' library(bemendel)
#' mat_geno <- split_geno("AaBb")$Loci # Get split loci and alleles
#' mat_geno
#'
#' @export
#'
split_geno <- function(x){

   geno <- gsub(" ", "", x) # Remove white spaces

   # Get the number of loci for diploid species
   if (nchar(geno) %% 2 == 0) {

   NLoci <- nchar(geno)/2

   } else {

     stop("Number of characters for inputted genotype must be a multiple of 2")

   }

  # Split inputted genotype into individual loci
  Loci <- strsplit(geno, "(?<=.{2})", perl = TRUE)

  # Split genotypes at individual loci into alleles
  Loci.ls <- strsplit(unlist(Loci), "")

  # Rename each locus using LETTERS
  names(Loci.ls) <- paste0("Locus_", 1:NLoci)

  out <- list(Loci = Loci.ls, NLoci = NLoci)

  return(out)

}


#' #' Check whether locus will segregate using this function
#' #' Returns a statement about whether locus is homo or hetero
#' #'
#' Seg <- function(x){
#'
#'   if (x[[1]] == x[[2]]) {
#'     paste("Locus", toupper(x[[1]]), "is homozygous, hence would not segregate.")
#'
#'   } else (paste("Locus", toupper(x[[1]]), "is heterozygous, hence would segregate."))
#' }
#'
#' # Test Seq func
#' #' Output is a string object
#' Seg(Loci.ls[[1]])
#' cat(unlist(sapply(Loci.ls, Seg)), "\n")



# Derive gametes for both parents
#' Output is a data frame object

#' Function to generate parental gametes
#'
#' @param x a list object containing split loci and alleles generated by
#' the `split_geno()` function
#'
#' @returns a factor object that contain all possible gametes that can be
#'  produced by an individual at meiosis.
#'
#'  @examples
#'
#'  library(bemendel)
#'  mat_geno <- split_geno("AaBb")$Loci # Get split loci and alleles
#'  mat_gam <- gamete(mat_geno)
#'  mat_gam
#'
#' @export
#'
gamete <- function(x){

  gg <- expand.grid(x)# Generates allele combinations

  # Find unique gametes; output is a factor object of unique gametes
  gam <- unique(interaction(gg[1:nrow(gg),], sep = ""))

  return(gam)

}



#' Function to order allele combinations at a locus -- Dominant allele first
#' Order alleles (dominant allele comes first at each locus)
#'
#' @param x a list object containing split loci and alleles at each locus
#'
#' @returns ordered alleles at each locus
#'
#' @examples
#'
#'  library(bemendel)
#'  mat_geno <- split_geno("AaBb")$Loci # Get split loci and alleles
#'  order_als(mat_geno$Locus_1)
#'
#' @export
#'
order_als <- function(x) {

  if (x[[1]] == x[[2]]) {

    paste0(x[[1]], x[[1]])

  } else if (x[[1]] != x[[2]] & toupper(x[[1]]) == toupper(x[[2]])) {

  paste0(toupper(x[[1]]), tolower(x[[1]]))

  } else if (x[[1]] != x[[2]] & toupper(x[[1]]) != toupper(x[[2]])) {

    paste0(x[[1]], x[[2]])}
}




#' Function to create punnett square given the genotypes of parents
#'
#' @param female.geno is a character object for the genotype of the female
#' parent
#' @param male.geno  is a character object for the genotype of the male parent
#'
#' @returns a punnett square table showing the genotypes of progenies in a
#' genetic cross
#'
#' @examples
#'
#' library(bemendel)
#' pun1 <- do_pun(female.geno = "AaBbDd", male.geno = "AaBbDd")
#' pun1$punnett_square # Get punnett square
#' pun1$pun_summary # Get punnett square summary
#'
#'
#' @export

do_pun <- function(female.geno, male.geno) {

  Loci.fem <-  split_geno(female.geno)$Loci # Loci and alleles in female parent
  Loci.mal <-  split_geno(male.geno)$Loci   # Loci and alleles in male parent

  # NLoci <- split_geno("AaBb")$NLoci

  fem.gam <-  gamete(Loci.fem) # Generate female gametes
  mal.gam <- gamete(Loci.mal) # Generate male gametes

# Create an empty matrix to hold genotypes of progenies after a cross
pun1 <- matrix(NA, nrow = length(fem.gam), ncol = length(mal.gam))
rownames(pun1) <- fem.gam # Assign female gametes
colnames(pun1) <- mal.gam # Assign male gametes

for (i in rownames(pun1)) {

  for (j in colnames(pun1)) {

    # Combine female and male gametes to generate progeny genotypes and organize
    # alleles occupying each locus
    Geno <- paste0(sort(unlist(strsplit(c(i, j), ""))),
                   collapse = "")

    # Split genotypes at each locus into alleles -- a list object output
    FF <- split_geno(Geno)$Loci

    # Re-order alleles at each locus and re-paste across all loci
    KK <- paste0(sapply(FF, order_als), collapse = "")

    pun1[i, j] <- KK

     }

  }

# Summaries of progeny genotypes in punnett square
# Convert genotype frequencies into fractions
Geno.freq <- MASS::fractions(rev(table(pun1))/sum(table(pun1)))

tab1 <- data.frame(Geno.freq)
colnames(tab1)[1] <- "Genotype"

if (dim(tab1)[1] > 1) {

  tab1$freq2 <- as.character(MASS::fractions(tab1$Freq))

} else (tab1$freq2 <- "1/1")


tab2 <- data.frame(pun1)

tabs <- list(punnett_square = tab2, pun_summary = tab1)

return(tabs)

} # End of do_pun() function


#' Function to melt punnett square to a long format data frame for plotting
#'
#' @param pun a matrix of punnett square containing progeny genotypes
#'
#' @returns a long format data frame for plotting punnett square
#'
#' @examples
#'
#' library(bemendel)
#' pun1 <- do_pun(female.geno = "AaBbDd", male.geno = "AaBbDd")
#' long_df <- pun2df(pun1$punnett_square)
#' long_df
#'
#' @export
#'
#'
pun2df <- function(pun){

  # Convert punnett square to a vector object and split progeny genotypes
  # into loci
  x <- strsplit(unname(unlist(as.vector(pun))), "(?<=.{2})", perl = TRUE)

  NLoci <- length(x[[1]]) # Get the number of loci

  # Loop to find the phenotypic class of each genotype
  # Assuming complete dominance at each locus and independent assortment
  for (i in seq_len(length(x))) {

    for (j in seq_len(NLoci)) {

      if (x[[i]][j] == tolower(x[[i]][j])) {

        x[[i]][j] <- x[[i]][j]

      } else {

        substr(x[[i]][j], start = 2, stop = 2) <- '_'

      }
    }

    # Re-paste split and reformatted progeny genotypes together
    x[[i]] <- paste0(x[[i]], collapse = "")

  }

  # Melt punnett square into a long format data frame for plotting using
  # ggplot2
  testdf <- reshape2::melt(as.matrix(pun), value.name = "geno")

  #' Add phenotypic group for each genotype to the melted data frame
  testdf$phenotype <- as.vector(unlist(x))

  return(testdf)

}




#' Function to check if locus is heterozygous in diploid species
#'
#' @param x a character vector of split alleles for each locus--
#' output of the `split_geno()` function.
#'
#' @returns a logical value of TRUE if locus is heterozygous or FALSE if locus is
#' homozygous.
#'
#' @examples
#'
#' library(bemendel)
#' is.het(split_geno("AaBb")$Loci[[1]])
#'
#' @export

is.het <- function(x){

  if (x[1] != x[2]) {

    het <- TRUE

  } else if (x[1] == x[2]) {

    het <- FALSE

  }

  return(het)

}

#' Function to display classical digenic epistatic ratios. Five types are
#' available
#'
#' @param female.geno is a character object for the genotype of the female
#' parent.
#'
#' @param male.geno  is a character object for the genotype of the male parent.
#'
#' @param type a string abbreviation indicating the type of digenic epistasis;
#' see details for more information.
#'
#' @returns a melted long format data frame.
#'
#' @examples
#'
#' library(bemendel)
#' epi_de <- epist(female.geno = "AaBb", male.geno = "AaBb", type = "DE")
#' epi_de
#' table(epi_de$epistasis)
#'
#'
#' @details
#' The following classical digenic epistasis types are implemented:
#' DDE: Duplicate dominant epistasis (15:1 ratio)
#' DRE: Duplicate recessive epistasis (9:7 ratio)
#' DE: Dominant epistasis (12:3:1 ratio)
#' RE: Recessive epistasis (9:3:4 ratio)
#' DnRE: Dominant and recessive epistasis (13:3 ratio)
#'
#' @export
#'
epist <- function(female.geno,
                  male.geno,
                  type = c('DDE', 'DRE', 'DE', 'RE', 'DnRE')){

  # Check if loci are heterozygous
  mm <- all(sapply(split_geno(female.geno)$Loci, is.het))
  pp <- all(sapply(split_geno(female.geno)$Loci, is.het))

  # Number of gene loci
  NLoci.m <- split_geno(female.geno)$NLoci
  NLoci.p <- split_geno(male.geno)$NLoci

  if (NLoci.m != 2 || NLoci.m != 2) {

    stop("Number of loci for each parent = 2")

  }

  if (mm != TRUE || pp != TRUE) {


    stop("Both parents must be dihybrids.")

  }


  pun <- do_pun(female.geno = female.geno, male.geno = male.geno)$punnett_square

  x <- pun2df(pun = pun) # convert punnett to long format data frame

  # Extract unique phenotypes under independent assortment
  uu <- sort(unique(x$phenotype), decreasing = FALSE)

  # Re-code phenotypes based on the type of epistasis

  epi <-  switch(type,
                 DE = ifelse(x$phenotype == uu[1] | x$phenotype == uu[2],
                             paste(uu[1], uu[2], sep = '||'),
                             ifelse(x$phenotype == uu[3],
                                    paste(uu[3]), paste(uu[4]))),

                 RE = ifelse(x$phenotype == uu[1], paste(uu[1]),
                             ifelse(x$phenotype == uu[3] | x$phenotype == uu[4],
                                    paste(uu[3], uu[4], sep = '||'), paste(uu[2]))),

                 DDE = ifelse(x$phenotype == uu[1] | x$phenotype == uu[2] |
                                x$phenotype == uu[3], paste(uu[1], uu[2], uu[3], sep = '||'),
                              paste(uu[4])),

                 DRE = ifelse(x$phenotype == uu[2] | x$phenotype == uu[3] |
                                x$phenotype == uu[4], paste(uu[2], uu[3], uu[4], sep = '||'),
                              paste(uu[1])),

                 DnRE = ifelse(x$phenotype == uu[1] | x$phenotype == uu[2] |
                                 x$phenotype == uu[4], paste(uu[1], uu[2], uu[4], sep = '||'),
                               paste(uu[3])),

                 stop("Invalid `epistasis type` value")

  )

  x$epistasis <- epi # Add epistatic phenotypes to melted data frame
  x$epi_type <- type # Add type of epistasis

  return(x)

}



#' Plot Punnett square to visualize phenotype classes of progenies assuming
#' complete dominance at each locus. The function helps users to visualize
#' predictions for common gene interaction hypotheses such as independent
#' assortment and classical digenic interactions.
#'
#' @param pun2df a melted long format data frame of punnett square containing
#' progeny genotypes and phenotypic class designation
#' @param text_size a numeric value for setting text size in plot
#' @param epistasis a logical value indicating whether input data frame contains
#' digenic epistatic phenotypes.
#'
#'
#' @examples
#'
#' library(bemendel)
#'
#' # Generate punnett square
#' pun1 <- do_pun(female.geno = "AaBb", male.geno = "AaBb")
#'
#' # Melt punnett to a long format data frame for plotting
#' long_df <- pun2df(pun1$punnett_square)
#'
#' # Visualize phenotypic groups based on independent assortment
#' pun_plot(pun2df = long_df, text_size = 6)
#'
#' # Visualize phenotypic groups based on dominant epistatic gene action
#' # Dominant epistasis (DE)
#' epi_de <- epist(female.geno = "AaBb", male.geno = "AaBb", type = "DE")
#' pun_plot(pun2df = epi_de, text_size = 6, epistasis = TRUE)
#'
#'
#' @details
#' Predictions of phenotypic classes for independent assortment are based on
#' complete dominance at all gene loci. For epistasis, predictions are based on
#' intergenic allelic interactions between two loci.
#'
#' For epistasis, the number of loci must be equal to 2. Digenic epistatic
#' ratios that can be visualized are
#' Duplicate dominant epistasis = 15:1 (Duplicate gene action)
#' Duplicate recessive epistasis = 9:7 (Complementary gene action)
#' Dominant epistasis = 12:3:1
#' Recessive epistasis = 9:3:4
#' Dominant and recessive epistasis = 13:3 (recessive suppressors)
#'
#' @returns a plot of the punnett square showing the phenotypic classes for
#'
#'
#' @export

pun_plot <- function(pun2df, text_size = 12, epistasis = FALSE) {

  Var1 <- Var2 <- geno <- phenotype <- NULL

  testdf <- pun2df

  NLoci <- nchar(testdf$geno[1])/2 # Number of loci

  # Get genotypic phenotypic ratio from punnett square
  if (epistasis == FALSE) {
  ratio <- table(testdf$phenotype)
  ratio <- paste(ratio, collapse = ":")

  # Plot title
  title <- paste('Independent assortment at all loci',
                   paste0('(', ratio, ')'))

  } else {

    testdf$phenotype <- testdf$epistasis
    ratio <- table(testdf$phenotype)
    ratio <- paste(ratio, collapse = ":")

    type <- testdf$epi_type[1]

    title <- switch(type,
                    DE = paste('Dominant epistasis', paste0('(', ratio, ')')),

                    RE = paste('Recessive epistasis', paste0('(', ratio, ')')),

                    DDE = paste('Duplicate dominant epistasis', paste0('(', ratio, ')')),

                    DRE = paste('Duplicate recessive epistasis', paste0('(', ratio, ')')),

                    DnRE = paste('Dominant and recessive epistasis', paste0('(', ratio, ')')),

                    stop("Invalid `epistasis type` value")

    )

  }

  # testdf <- testdf[order(testdf$phenotype, decreasing = TRUE),]

  plt <- ggplot2::ggplot(testdf, ggplot2::aes(x = Var1, y = Var2,
                                            label = geno, fill = phenotype)) +
    ggplot2::theme(axis.text = ggplot2::element_text(size  = text_size/NLoci),
                   axis.title = ggplot2::element_text(size = text_size, face = "bold")) +
    ggplot2::labs(x = 'Female gamete', y = 'Male gamete', fill = 'Phenotype',
                  title = title,
                  subtitle = paste('This cross involved', NLoci, 'gene loci.')) +
    ggplot2::theme_classic() +
    ggplot2::theme(plot.title = ggplot2::element_text(family = "serif",
                                                      color = "grey10",
                                                      size = 12,
                                                      hjust = 0.5),
                   plot.subtitle = ggplot2::element_text(hjust = 0,
                                                         family = "serif",
                                                         face = "bold",
                                                         color = "grey20",
                                                         size = 10),
                   legend.title = ggplot2::element_text(color = "black", size = text_size),
                   legend.text = ggplot2::element_text(size = text_size, colour = "blue4"),
                   axis.line = ggplot2::element_blank()) +
    ggplot2::geom_text(colour = "black", size = text_size/NLoci, fontface = 'bold',
                       hjust = 0.5) +
    ggplot2::geom_tile(alpha = 0.5)

  return(plt)

}





